diff --git a/qiling/loader/elf.py b/qiling/loader/elf.py
index a5a8ac6f..56d91a8e 100644
--- a/qiling/loader/elf.py
+++ b/qiling/loader/elf.py
@@ -59,6 +59,8 @@ class AUXV(IntEnum):
 # we will reserve 0x1000 bytes for this (which contains multiple slots of 4/8 bytes, each for one api)
 API_HOOK_MEM = 0x1000000
 API_HOOK_SIZE = 0x1000
+# cleanup memory layout
+HOOK_MEM = 0xeeeee000
 
 # memory for syscall table
 SYSCALL_MEM = API_HOOK_MEM + 0x1000
@@ -100,8 +102,12 @@ class QlLoaderELF(QlLoader):
 
         # is it a driver?
         if elftype == 'ET_REL':
-            self.load_driver(elffile, top_of_stack, loadbase=0x8000000)
-            self.ql.hook_code(hook_kernel_api)
+            # why the fuck is the TC TA ET_REL????
+            load_address = 0
+            self.stack_address = stack_address+0x3000
+            self.load_address = load_address
+            #self.load_driver(elffile, top_of_stack, loadbase=0x8000000)
+            #self.ql.hook_code(hook_kernel_api)
 
         # is it an executable?
         elif elftype == 'ET_EXEC':
@@ -189,7 +195,7 @@ class QlLoaderELF(QlLoader):
                         #
                         # this workaround unifies such "overlapping" segments, which may apply more permissive
                         # protection flags to that memory region.
-                        if self.ql.arch.type == QL_ARCH.ARM64:
+                        if self.ql.arch.type == QL_ARCH.ARM64 or self.ql.arch.type == QL_ARCH.ARM:
                             load_regions[-1] = (prev_lbound, ubound, prev_perms | perms)
                             continue
 
@@ -387,7 +393,7 @@ class QlLoaderELF(QlLoader):
         self.init_sp = self.ql.arch.regs.arch_sp
 
         self.ql.os.entry_point = self.entry_point = entry_point
-        self.ql.os.function_hook = FunctionHook(self.ql, elf_phdr, elf_phnum, elf_phent, load_address, mem_end)
+        self.ql.os.function_hook = FunctionHook(self.ql, elf_phdr, elf_phnum, elf_phent, load_address, HOOK_MEM)
 
         # If there is a loader, we ignore exit
         self.skip_exit_check = (self.elf_entry != self.entry_point)
diff --git a/qiling/log.py b/qiling/log.py
index 2735b8b7..0774a810 100644
--- a/qiling/log.py
+++ b/qiling/log.py
@@ -193,8 +193,9 @@ def setup_logger(ql: Qiling, logdevs: Collection[Union[IO, str]], plain: bool, o
 
             elif isinstance(dev, str):
                 handler = FileHandler(dev)
-
+            
             else:
+                #handler = StreamHandler(dev)
                 raise TypeError(f'unexpected logging device type: {type(dev).__name__}')
 
             if no_color or not __is_color_terminal(handler.stream):
